File: software/app/js/protocol.js
// 本文件由 protocol/tools/gen.py 从 contract.yaml 生成，请勿手改。
// contract version: 0.2.0-demo

export const NlConst = Object.freeze({
  protoVersion: "0.2.0-demo",
  protoMagic: 20026,
  versionMajor: 0,
  versionMinor: 2,
  levelMin: 1,
  levelMax: 8,
  dutyCapPct: 90,
  uplinkHz: 12,
  uplinkPeriodMs: 83,
  ledCount: 8,
  dht11MinIntervalMs: 1000,
  imuDecisionPeriodMs: 1000,
  stillPauseMs: 30000,
  joyEdgeHoldMs: 80,
  joyHoldRampMs: 400,
  joyDeadzone: 900,
  wildTimeoutMs: 900000,
  linkTimeoutMs: 1500,
  sessionTokenTtlMs: 3600000,
  sentinelI16: -32768,
});

export const NlBle = Object.freeze({
  deviceName: "Nascent-K10",
  minMtu: 185,
  serviceUuid: "a1b2c000-5f3e-4c8a-9b1d-0e7f2a6c9d10",
  uplinkUuid: "a1b2c001-5f3e-4c8a-9b1d-0e7f2a6c9d10",
  downlinkUuid: "a1b2c002-5f3e-4c8a-9b1d-0e7f2a6c9d10",
  infoUuid: "a1b2c003-5f3e-4c8a-9b1d-0e7f2a6c9d10",
});

export const NlFrameType = Object.freeze((() => {
  const values = ["pair", "ack", "heartbeat", "telemetry", "command"];
  return {
    PAIR: "pair",
    ACK: "ack",
    HEARTBEAT: "heartbeat",
    TELEMETRY: "telemetry",
    COMMAND: "command",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlMode = Object.freeze((() => {
  const values = ["free", "scenario", "wild"];
  return {
    FREE: "free",
    SCENARIO: "scenario",
    WILD: "wild",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlInsertState = Object.freeze((() => {
  const values = ["unknown", "not_inserted", "inserted"];
  return {
    UNKNOWN: "unknown",
    NOT_INSERTED: "not_inserted",
    INSERTED: "inserted",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlJoyEdge = Object.freeze((() => {
  const values = ["none", "up", "down"];
  return {
    NONE: "none",
    UP: "up",
    DOWN: "down",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlAlert = Object.freeze((() => {
  const values = ["none", "over_temp", "low_battery", "safeword", "estop", "bad_cmd", "link_lost"];
  return {
    NONE: "none",
    OVER_TEMP: "over_temp",
    LOW_BATTERY: "low_battery",
    SAFEWORD: "safeword",
    ESTOP: "estop",
    BAD_CMD: "bad_cmd",
    LINK_LOST: "link_lost",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlCmd = Object.freeze((() => {
  const values = ["stop", "set_mode", "set_level", "set_pattern", "set_led", "set_joystick", "resume"];
  return {
    STOP: "stop",
    SET_MODE: "set_mode",
    SET_LEVEL: "set_level",
    SET_PATTERN: "set_pattern",
    SET_LED: "set_led",
    SET_JOYSTICK: "set_joystick",
    RESUME: "resume",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlPattern = Object.freeze((() => {
  const values = ["soft_min", "soft", "shallow_wave", "wave", "pulse", "strong_pulse", "mixed", "peak"];
  return {
    SOFT_MIN: "soft_min",
    SOFT: "soft",
    SHALLOW_WAVE: "shallow_wave",
    WAVE: "wave",
    PULSE: "pulse",
    STRONG_PULSE: "strong_pulse",
    MIXED: "mixed",
    PEAK: "peak",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlLedState = Object.freeze((() => {
  const values = ["mode_default", "warming", "comfort_reached", "cleaning", "low_battery", "safeword"];
  return {
    MODE_DEFAULT: "mode_default",
    WARMING: "warming",
    COMFORT_REACHED: "comfort_reached",
    CLEANING: "cleaning",
    LOW_BATTERY: "low_battery",
    SAFEWORD: "safeword",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlPhase = Object.freeze((() => {
  const values = ["warming", "rising", "plateau", "calm"];
  return {
    WARMING: "warming",
    RISING: "rising",
    PLATEAU: "plateau",
    CALM: "calm",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlSceneCtrl = Object.freeze((() => {
  const values = ["advance", "stay", "end"];
  return {
    ADVANCE: "advance",
    STAY: "stay",
    END: "end",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlEmotion = Object.freeze((() => {
  const values = ["gentle", "playful", "calm"];
  return {
    GENTLE: "gentle",
    PLAYFUL: "playful",
    CALM: "calm",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlMoodTone = Object.freeze((() => {
  const values = ["quiet", "open", "warm", "bright", "tired"];
  return {
    QUIET: "quiet",
    OPEN: "open",
    WARM: "warm",
    BRIGHT: "bright",
    TIRED: "tired",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlTempState = Object.freeze((() => {
  const values = ["unknown", "too_cold", "warming", "reaching_comfort", "comfortable", "cooling"];
  return {
    UNKNOWN: "unknown",
    TOO_COLD: "too_cold",
    WARMING: "warming",
    REACHING_COMFORT: "reaching_comfort",
    COMFORTABLE: "comfortable",
    COOLING: "cooling",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const NlRhythm = Object.freeze((() => {
  const values = ["unknown", "steady", "increasing", "decreasing"];
  return {
    UNKNOWN: "unknown",
    STEADY: "steady",
    INCREASING: "increasing",
    DECREASING: "decreasing",
    values,
    fromWire(i) { return (i >= 0 && i < values.length) ? values[i] : values[0]; },
    fromWireName(n) {
      const i = values.indexOf(n ?? '');
      return i < 0 ? values[0] : values[i];
    },
  };
})());

export const kLevelTable = Object.freeze([
  Object.freeze({ level: 1, dutyPct: 15, pattern: NlPattern.SOFT_MIN, lit: 1, r: 255, g: 105, b: 150, semantic: "初识预热" }),
  Object.freeze({ level: 2, dutyPct: 25, pattern: NlPattern.SOFT, lit: 2, r: 255, g: 105, b: 150, semantic: "轻抚" }),
  Object.freeze({ level: 3, dutyPct: 35, pattern: NlPattern.SHALLOW_WAVE, lit: 3, r: 255, g: 130, b: 110, semantic: "渐入" }),
  Object.freeze({ level: 4, dutyPct: 45, pattern: NlPattern.WAVE, lit: 4, r: 255, g: 130, b: 110, semantic: "升温" }),
  Object.freeze({ level: 5, dutyPct: 55, pattern: NlPattern.PULSE, lit: 5, r: 255, g: 150, b: 60, semantic: "攀升" }),
  Object.freeze({ level: 6, dutyPct: 65, pattern: NlPattern.STRONG_PULSE, lit: 6, r: 255, g: 110, b: 50, semantic: "深入" }),
  Object.freeze({ level: 7, dutyPct: 78, pattern: NlPattern.MIXED, lit: 7, r: 255, g: 70, b: 60, semantic: "高强度" }),
  Object.freeze({ level: 8, dutyPct: 90, pattern: NlPattern.PEAK, lit: 8, r: 255, g: 70, b: 60, semantic: "峰值" }),
]);

export class LevelSetting {
  constructor({
    level,
    pattern,
  }) {
    this.level = level;
    this.pattern = pattern;
  }

  static fromJson(j) {
    return new LevelSetting({
      level: Number(j["level"]),
      pattern: NlPattern.fromWireName(j["pattern"]),
    });
  }

  toJson() {
    return {
      "level": this.level,
      "pattern": this.pattern,
    };
  }
}

export class CloudAction {
  constructor({
    setLevel = null,
    setPattern = null,
  }) {
    this.setLevel = setLevel;
    this.setPattern = setPattern;
  }

  static fromJson(j) {
    return new CloudAction({
      setLevel: j["set_level"] == null ? null : Number(j["set_level"]),
      setPattern: j["set_pattern"] == null ? null : NlPattern.fromWireName(j["set_pattern"]),
    });
  }

  toJson() {
    return {
      "set_level": this.setLevel,
      "set_pattern": this.setPattern,
    };
  }
}

export class SessionCurves {
  constructor({
    tempA1hz,
    press1hz,
    hr1hz,
  }) {
    this.tempA1hz = tempA1hz;
    this.press1hz = press1hz;
    this.hr1hz = hr1hz;
  }

  static fromJson(j) {
    return new SessionCurves({
      tempA1hz: (j["temp_a_1hz"]).map((e) => Number(e)),
      press1hz: (j["press_1hz"]).map((e) => Number(e)),
      hr1hz: (j["hr_1hz"]).map((e) => Number(e)),
    });
  }

  toJson() {
    return {
      "temp_a_1hz": this.tempA1hz,
      "press_1hz": this.press1hz,
      "hr_1hz": this.hr1hz,
    };
  }
}

export class SessionEvent {
  constructor({
    t,
    type,
  }) {
    this.t = t;
    this.type = type;
  }

  static fromJson(j) {
    return new SessionEvent({
      t: Number(j["t"]),
      type: String(j["type"]),
    });
  }

  toJson() {
    return {
      "t": this.t,
      "type": this.type,
    };
  }
}

export class UserTags {
  constructor({
    mood,
    note,
  }) {
    this.mood = mood;
    this.note = note;
  }

  static fromJson(j) {
    return new UserTags({
      mood: NlMoodTone.fromWireName(j["mood"]),
      note: String(j["note"]),
    });
  }

  toJson() {
    return {
      "mood": this.mood,
      "note": this.note,
    };
  }
}

/** k10-controller -> app（BLE notify，12 Hz 聚合） */
export class BleUplink {
  constructor({
    ts,
    tempA = null,
    tempB = null,
    envTemp = null,
    envHumidity = null,
    pressL,
    pressR,
    accel,
    gyro,
    insertState,
    joyEdge,
    mode,
    level,
    battery = null,
    alert,
  }) {
    this.ts = ts;
    this.tempA = tempA;
    this.tempB = tempB;
    this.envTemp = envTemp;
    this.envHumidity = envHumidity;
    this.pressL = pressL;
    this.pressR = pressR;
    this.accel = accel;
    this.gyro = gyro;
    this.insertState = insertState;
    this.joyEdge = joyEdge;
    this.mode = mode;
    this.level = level;
    this.battery = battery;
    this.alert = alert;
  }

  static fromJson(j) {
    return new BleUplink({
      ts: Number(j["ts"]),
      // 量产接触 NTC；demo 恒为 null
      tempA: j["temp_a"] == null ? null : Number(j["temp_a"]),
      // 量产环境 NTC；demo 恒为 null
      tempB: j["temp_b"] == null ? null : Number(j["temp_b"]),
      // DHT11 温度，最快 1 Hz，非安全通道
      envTemp: j["env_temp"] == null ? null : Number(j["env_temp"]),
      // DHT11 湿度，最快 1 Hz
      envHumidity: j["env_humidity"] == null ? null : Number(j["env_humidity"]),
      pressL: Number(j["press_l"]),
      pressR: Number(j["press_r"]),
      accel: (j["accel"]).map((e) => Number(e)),
      gyro: (j["gyro"]).map((e) => Number(e)),
      insertState: NlInsertState.fromWireName(j["insert_state"]),
      joyEdge: NlJoyEdge.fromWireName(j["joy_edge"]),
      mode: NlMode.fromWireName(j["mode"]),
      level: Number(j["level"]),
      // demo 恒为 null
      battery: j["battery"] == null ? null : Number(j["battery"]),
      alert: NlAlert.fromWireName(j["alert"]),
    });
  }

  toJson() {
    return {
      "ts": this.ts,
      "temp_a": this.tempA,
      "temp_b": this.tempB,
      "env_temp": this.envTemp,
      "env_humidity": this.envHumidity,
      "press_l": this.pressL,
      "press_r": this.pressR,
      "accel": this.accel,
      "gyro": this.gyro,
      "insert_state": this.insertState,
      "joy_edge": this.joyEdge,
      "mode": this.mode,
      "level": this.level,
      "battery": this.battery,
      "alert": this.alert,
    };
  }
}

/** app -> k10-controller（BLE write） */
export class BleDownlink {
  constructor({
    cmd,
    level = null,
    pattern = null,
    mode = null,
    led = null,
    enabled = null,
    holdRamp = null,
    auth,
  }) {
    this.cmd = cmd;
    this.level = level;
    this.pattern = pattern;
    this.mode = mode;
    this.led = led;
    this.enabled = enabled;
    this.holdRamp = holdRamp;
    this.auth = auth;
  }

  static fromJson(j) {
    return new BleDownlink({
      cmd: NlCmd.fromWireName(j["cmd"]),
      level: j["level"] == null ? null : Number(j["level"]),
      pattern: j["pattern"] == null ? null : NlPattern.fromWireName(j["pattern"]),
      mode: j["mode"] == null ? null : NlMode.fromWireName(j["mode"]),
      led: j["led"] == null ? null : NlLedState.fromWireName(j["led"]),
      // set_joystick 用
      enabled: j["enabled"] == null ? null : Boolean(j["enabled"]),
      // set_joystick 用
      holdRamp: j["hold_ramp"] == null ? null : Boolean(j["hold_ramp"]),
      // session token；缺失或过期整包丢弃
      auth: String(j["auth"]),
    });
  }

  toJson() {
    return {
      "cmd": this.cmd,
      "level": this.level,
      "pattern": this.pattern,
      "mode": this.mode,
      "led": this.led,
      "enabled": this.enabled,
      "hold_ramp": this.holdRamp,
      "auth": this.auth,
    };
  }
}

/** app -> backend（5-10s 或事件触发） */
export class CloudSummary {
  constructor({
    sessionId,
    personaId,
    mode,
    phase,
    tempState,
    pressureRhythm,
    hrTrend,
    insertState,
    userEvents,
    current,
  }) {
    this.sessionId = sessionId;
    this.personaId = personaId;
    this.mode = mode;
    this.phase = phase;
    this.tempState = tempState;
    this.pressureRhythm = pressureRhythm;
    this.hrTrend = hrTrend;
    this.insertState = insertState;
    this.userEvents = userEvents;
    this.current = current;
  }

  static fromJson(j) {
    return new CloudSummary({
      sessionId: String(j["session_id"]),
      personaId: String(j["persona_id"]),
      mode: NlMode.fromWireName(j["mode"]),
      phase: NlPhase.fromWireName(j["phase"]),
      tempState: NlTempState.fromWireName(j["temp_state"]),
      pressureRhythm: NlRhythm.fromWireName(j["pressure_rhythm"]),
      hrTrend: NlRhythm.fromWireName(j["hr_trend"]),
      insertState: NlInsertState.fromWireName(j["insert_state"]),
      userEvents: (j["user_events"]).map((e) => String(e)),
      current: LevelSetting.fromJson(j["current"]),
    });
  }

  toJson() {
    return {
      "session_id": this.sessionId,
      "persona_id": this.personaId,
      "mode": this.mode,
      "phase": this.phase,
      "temp_state": this.tempState,
      "pressure_rhythm": this.pressureRhythm,
      "hr_trend": this.hrTrend,
      "insert_state": this.insertState,
      "user_events": this.userEvents,
      "current": this.current.toJson(),
    };
  }
}

/** backend -> app */
export class CloudActionEnvelope {
  constructor({
    dialogue,
    action = null,
    sceneCtrl,
    emotion,
  }) {
    this.dialogue = dialogue;
    this.action = action;
    this.sceneCtrl = sceneCtrl;
    this.emotion = emotion;
  }

  static fromJson(j) {
    return new CloudActionEnvelope({
      dialogue: String(j["dialogue"]),
      action: j["action"] == null ? null : CloudAction.fromJson(j["action"]),
      sceneCtrl: NlSceneCtrl.fromWireName(j["scene_ctrl"]),
      emotion: NlEmotion.fromWireName(j["emotion"]),
    });
  }

  toJson() {
    return {
      "dialogue": this.dialogue,
      "action": this.action?.toJson(),
      "scene_ctrl": this.sceneCtrl,
      "emotion": this.emotion,
    };
  }
}

/** app 本地 SQLCipher 归档 */
export class SessionRecord {
  constructor({
    sessionId,
    tsStart,
    durationS,
    mode,
    personaId,
    sceneId = null,
    maxLevel,
    curves,
    events,
    userTags,
    aiSummaryDraft,
  }) {
    this.sessionId = sessionId;
    this.tsStart = tsStart;
    this.durationS = durationS;
    this.mode = mode;
    this.personaId = personaId;
    this.sceneId = sceneId;
    this.maxLevel = maxLevel;
    this.curves = curves;
    this.events = events;
    this.userTags = userTags;
    this.aiSummaryDraft = aiSummaryDraft;
  }

  static fromJson(j) {
    return new SessionRecord({
      sessionId: String(j["session_id"]),
      tsStart: Number(j["ts_start"]),
      durationS: Number(j["duration_s"]),
      mode: NlMode.fromWireName(j["mode"]),
      personaId: String(j["persona_id"]),
      sceneId: j["scene_id"] == null ? null : String(j["scene_id"]),
      maxLevel: Number(j["max_level"]),
      curves: SessionCurves.fromJson(j["curves"]),
      events: (j["events"]).map((e) => SessionEvent.fromJson(e)),
      userTags: UserTags.fromJson(j["user_tags"]),
      aiSummaryDraft: String(j["ai_summary_draft"]),
    });
  }

  toJson() {
    return {
      "session_id": this.sessionId,
      "ts_start": this.tsStart,
      "duration_s": this.durationS,
      "mode": this.mode,
      "persona_id": this.personaId,
      "scene_id": this.sceneId,
      "max_level": this.maxLevel,
      "curves": this.curves.toJson(),
      "events": this.events.map((e) => e.toJson()),
      "user_tags": this.userTags.toJson(),
      "ai_summary_draft": this.aiSummaryDraft,
    };
  }
}

File: software/backend/app/protocol.py
"""本文件由 protocol/tools/gen.py 从 contract.yaml 生成，请勿手改。

contract version: 0.3.0-demo
"""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel

PROTO_VERSION = "0.3.0-demo"


class NlConst:
    PROTO_MAGIC = 20026
    VERSION_MAJOR = 0
    VERSION_MINOR = 3
    LEVEL_MIN = 1
    LEVEL_MAX = 8
    DUTY_CAP_PCT = 90
    UPLINK_HZ = 12
    UPLINK_PERIOD_MS = 83
    LED_COUNT = 8
    DHT11_MIN_INTERVAL_MS = 1000
    IMU_DECISION_PERIOD_MS = 1000
    STILL_PAUSE_MS = 30000
    BOOT_KEY_DEBOUNCE_MS = 40
    BOOT_STOP_MAX_MS = 600
    BOOT_RESUME_HOLD_MS = 2000
    WILD_TIMEOUT_MS = 900000
    LINK_TIMEOUT_MS = 1500
    SESSION_TOKEN_TTL_MS = 3600000
    TRANSPORT_IDLE_SWITCH_MS = 20000
    SENTINEL_I16 = -32768


class Ble:
    """BLE GATT 标识，固件与 App 共用。"""

    DEVICE_NAME = "Nascent-Toy"
    MIN_MTU = 185
    SERVICE_UUID = "a1b2c000-5f3e-4c8a-9b1d-0e7f2a6c9d10"
    UPLINK_UUID = "a1b2c001-5f3e-4c8a-9b1d-0e7f2a6c9d10"
    DOWNLINK_UUID = "a1b2c002-5f3e-4c8a-9b1d-0e7f2a6c9d10"
    INFO_UUID = "a1b2c003-5f3e-4c8a-9b1d-0e7f2a6c9d10"


class Wifi:
    """WiFi 备用通道的端口、路径与 mDNS 主机名。"""

    WS_PORT = 81
    WS_PATH = "/nl"
    MDNS_HOST = "nascent"


class FrameType(str, Enum):
    TELEMETRY = "telemetry"
    COMMAND = "command"

    @property
    def wire(self) -> int:
        return list(FrameType).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "FrameType":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Mode(str, Enum):
    FREE = "free"
    SCENARIO = "scenario"
    WILD = "wild"

    @property
    def wire(self) -> int:
        return list(Mode).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Mode":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class InsertState(str, Enum):
    UNKNOWN = "unknown"
    NOT_INSERTED = "not_inserted"
    INSERTED = "inserted"

    @property
    def wire(self) -> int:
        return list(InsertState).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "InsertState":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Alert(str, Enum):
    NONE = "none"
    OVER_TEMP = "over_temp"
    LOW_BATTERY = "low_battery"
    SAFEWORD = "safeword"
    ESTOP = "estop"
    BAD_CMD = "bad_cmd"
    LINK_LOST = "link_lost"

    @property
    def wire(self) -> int:
        return list(Alert).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Alert":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Cmd(str, Enum):
    STOP = "stop"
    SET_MODE = "set_mode"
    SET_LEVEL = "set_level"
    SET_PATTERN = "set_pattern"
    SET_LED = "set_led"
    RESUME = "resume"
    SET_WIFI = "set_wifi"
    PRESS_KEY = "press_key"

    @property
    def wire(self) -> int:
        return list(Cmd).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Cmd":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class KeyPress(str, Enum):
    TAP = "tap"
    HOLD = "hold"

    @property
    def wire(self) -> int:
        return list(KeyPress).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "KeyPress":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Pattern(str, Enum):
    SOFT_MIN = "soft_min"
    SOFT = "soft"
    SHALLOW_WAVE = "shallow_wave"
    WAVE = "wave"
    PULSE = "pulse"
    STRONG_PULSE = "strong_pulse"
    MIXED = "mixed"
    PEAK = "peak"

    @property
    def wire(self) -> int:
        return list(Pattern).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Pattern":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class LedState(str, Enum):
    MODE_DEFAULT = "mode_default"
    WARMING = "warming"
    COMFORT_REACHED = "comfort_reached"
    CLEANING = "cleaning"
    LOW_BATTERY = "low_battery"
    SAFEWORD = "safeword"

    @property
    def wire(self) -> int:
        return list(LedState).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "LedState":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Phase(str, Enum):
    WARMING = "warming"
    RISING = "rising"
    PLATEAU = "plateau"
    CALM = "calm"

    @property
    def wire(self) -> int:
        return list(Phase).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Phase":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class SceneCtrl(str, Enum):
    ADVANCE = "advance"
    STAY = "stay"
    END = "end"

    @property
    def wire(self) -> int:
        return list(SceneCtrl).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "SceneCtrl":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Emotion(str, Enum):
    GENTLE = "gentle"
    PLAYFUL = "playful"
    CALM = "calm"

    @property
    def wire(self) -> int:
        return list(Emotion).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Emotion":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class MoodTone(str, Enum):
    QUIET = "quiet"
    OPEN = "open"
    WARM = "warm"
    BRIGHT = "bright"
    TIRED = "tired"

    @property
    def wire(self) -> int:
        return list(MoodTone).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "MoodTone":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class TempState(str, Enum):
    UNKNOWN = "unknown"
    TOO_COLD = "too_cold"
    WARMING = "warming"
    REACHING_COMFORT = "reaching_comfort"
    COMFORTABLE = "comfortable"
    COOLING = "cooling"

    @property
    def wire(self) -> int:
        return list(TempState).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "TempState":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


class Rhythm(str, Enum):
    UNKNOWN = "unknown"
    STEADY = "steady"
    INCREASING = "increasing"
    DECREASING = "decreasing"

    @property
    def wire(self) -> int:
        return list(Rhythm).index(self)

    @classmethod
    def from_wire(cls, i: int) -> "Rhythm":
        members = list(cls)
        return members[i] if 0 <= i < len(members) else members[0]


LEVEL_TABLE = (
    {"level": 1, "duty_pct": 15, "pattern": Pattern.SOFT_MIN, "lit": 1, "rgb": (255, 105, 150), "semantic": "初识预热"},
    {"level": 2, "duty_pct": 25, "pattern": Pattern.SOFT, "lit": 2, "rgb": (255, 105, 150), "semantic": "轻抚"},
    {"level": 3, "duty_pct": 35, "pattern": Pattern.SHALLOW_WAVE, "lit": 3, "rgb": (255, 130, 110), "semantic": "渐入"},
    {"level": 4, "duty_pct": 45, "pattern": Pattern.WAVE, "lit": 4, "rgb": (255, 130, 110), "semantic": "升温"},
    {"level": 5, "duty_pct": 55, "pattern": Pattern.PULSE, "lit": 5, "rgb": (255, 150, 60), "semantic": "攀升"},
    {"level": 6, "duty_pct": 65, "pattern": Pattern.STRONG_PULSE, "lit": 6, "rgb": (255, 110, 50), "semantic": "深入"},
    {"level": 7, "duty_pct": 78, "pattern": Pattern.MIXED, "lit": 7, "rgb": (255, 70, 60), "semantic": "高强度"},
    {"level": 8, "duty_pct": 90, "pattern": Pattern.PEAK, "lit": 8, "rgb": (255, 70, 60), "semantic": "峰值"},
)


class LevelSetting(BaseModel):
    level: int
    pattern: Pattern


class CloudAction(BaseModel):
    set_level: int | None = None
    set_pattern: Pattern | None = None


class SessionCurves(BaseModel):
    temp_a_1hz: list[float]
    press_1hz: list[float]
    hr_1hz: list[float]


class SessionEvent(BaseModel):
    t: int
    type: str


class UserTags(BaseModel):
    mood: MoodTone
    note: str


class BleUplink(BaseModel):
    """toy-sidecar -> app（BLE notify 或 WiFi WebSocket，12 Hz 聚合）"""

    ts: int
    temp_a: float | None = None  # 量产接触 NTC；demo 恒为 null
    temp_b: float | None = None  # 量产环境 NTC；demo 恒为 null
    env_temp: float | None = None  # DHT11 温度，最快 1 Hz，非安全通道
    env_humidity: float | None = None  # DHT11 湿度，最快 1 Hz
    press_l: int
    press_r: int
    accel: list[float]
    gyro: list[float]
    insert_state: InsertState
    mode: Mode
    level: int
    battery: int | None = None  # demo 恒为 null
    alert: Alert


class BleDownlink(BaseModel):
    """app -> toy-sidecar（BLE write 或 WiFi WebSocket）"""

    cmd: Cmd
    level: int | None = None
    pattern: Pattern | None = None
    mode: Mode | None = None
    led: LedState | None = None
    wifi_ssid: str | None = None  # 仅 set_wifi；写入玩具 NVS，不上云、不上行
    wifi_psk: str | None = None  # 仅 set_wifi；明文只走已鉴权的设备链路，固件不得打日志
    key: KeyPress | None = None  # 仅 press_key：tap 点按切档，hold 长按约 1.2s 开关机
    auth: str  # session token；缺失或过期整包丢弃


class CloudSummary(BaseModel):
    """app -> backend（5-10s 或事件触发）"""

    session_id: str
    persona_id: str
    mode: Mode
    phase: Phase
    temp_state: TempState
    pressure_rhythm: Rhythm
    hr_trend: Rhythm
    insert_state: InsertState
    user_events: list[str]
    current: LevelSetting


class CloudActionEnvelope(BaseModel):
    """backend -> app"""

    dialogue: str
    action: CloudAction | None = None
    scene_ctrl: SceneCtrl
    emotion: Emotion


class SessionRecord(BaseModel):
    """app 本地 SQLCipher 归档"""

    session_id: str
    ts_start: int
    duration_s: int
    mode: Mode
    persona_id: str
    scene_id: str | None = None
    max_level: int
    curves: SessionCurves
    events: list[SessionEvent]
    user_tags: UserTags
    ai_summary_draft: str
